pkg/table-rule-selector/trie_selector.go

```diff
@@ -0,0 +1,461 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package selector
+
+import (
+	"fmt"
+	"sync"
+
+	"github.com/juju/errors"
+)
+
+// 1. asterisk character (*, also called "star") matches zero or more characters,
+//    for example, doc* matches doc and document but not dodo;
+//    asterisk character must be the last character of wildcard word.
+// 2. the question mark ? matches exactly one character
+const (
+	// asterisk [ * ]
+	asterisk = '*'
+	// question mark [ ? ]
+	question = '?'
+)
+
+const maxCacheNum = 1024
+
+// Selector stores rules of schema/table for easy retrieval
+type Selector interface {
+	// Insert will insert one rule into trie
+	// if table is empty, insert rule into schema level
+	// otherwise insert rule into table level
+	Insert(schema, table string, rule interface{}, replace bool) error
+	// Match will return all matched rules
+	Match(schema, table string) RuleSet
+	// Remove will remove one rule
+	Remove(schema, table string) error
+	// AllRules will returns all rules
+	AllRules() (map[string]interface{}, map[string]map[string]interface{})
+}
+
+// RuleSet is a set of rules that selected
+type RuleSet []interface{}
+
+func (r RuleSet) clone() RuleSet {
+	if r == nil {
+		return nil
+	}
+
+	c := make(RuleSet, 0, len(r))
+	return append(c, r...)
+}
+
+type macthedResult struct {
+	nodes []*node
+	rules RuleSet
+}
+
+func (r *macthedResult) empty() bool {
+	return r == nil || (len(r.nodes) == 0 && len(r.rules) == 0)
+}
+
+type trieSelector struct {
+	sync.RWMutex
+
+	cache map[string]RuleSet
+	root  *node
+}
+
+type node struct {
+	characters         map[byte]*item
```

Comment:
how about make `characters` as type of `map[rune]*item`.
MySQL identifier may contain unicode chars, ref [https://dev.mysql.com/doc/refman/5.7/en/identifiers.html](https://dev.mysql.com/doc/refman/5.7/en/identifiers.html)

Reply:
otherwise we should refactor the use of `for range`

Reply:
yep, had talked about it with @dantin.  now it's just designed for uint8(ascii code).  we may change it later together with  related logic in `go-mysql`, `syncer`